pkg/rbd/rbd_util.go

```diff
@@ -550,6 +550,26 @@ func createSnapshot(pOpts *rbdSnapshot, adminID string, credentials map[string]s
 	return nil
 }
 
+// flattenImage remove the reference from the child clone to the parent
+// snapshot (refer http://docs.ceph.com/docs/giant/rbd/rbd-snapshot/#flattening-a-cloned-image)
+func flattenImage(vol *rbdVolume, credentials map[string]string) error {
+	var output []byte
+
+	key, err := getKey(vol.AdminID, credentials)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Infof("rbd: flatten image %s using mon %s, pool %s", vol.RbdImageName, vol.Monitors, vol.Pool)
+	args := []string{"flatten", "--pool", vol.Pool, "--image", vol.RbdImageName, "--id", vol.AdminID, "-m", vol.Monitors, "--key=" + key}
+
+	output, err = execCommand("rbd", args)
+
+	if err != nil {
```

Comment:
This should be expected to fail if the image doesn't have a parent. Additionally, if the image has snapshots and does not have the 'deep-flatten' feature enabled, the snapshots themselves will still be linked to the parent image.

Reply:
@dillaman Thanks for the review,
do I need to explicitly enable `deep-flatten` feature during `clone` and `create` or is  it enabled by default

Reply:
You cannot enable it on existing images, it needs to have been enabled when the image was created. Support for this in krbd wasn't added until v5.1.

Reply:
@dillaman don't we need to enable this feature during `clone` (cloned image can be used later to snapshot and clone a PVC from it)

Reply:
@Madhu-1 negative -- ```deep-flatten``` isn't required for cloning, ```layering``` is the required feature.

Thinking holistically, we added the "rbd trash mv" tool to support cases where you cannot delete a parent image due to linked clones. In the Nautilus release, this is even automated for you via the ```rbd_move_parent_to_trash_on_remove``` config option (requires Mimic-or-later cluster). We are also adding automated background operation handling [1]. 

From an operator's point-of-view, we should never really expose the RBD implementation details for how snapshots and volume from snapshots (RBD clones) work on the backend. If you attempt to delete an image that has linked clones, move it to the trash for later cleanup. If you attempt to delete a snapshot with a linked clone, w/ clone v2 it just works. To support clone v1 (and v2), perhaps k8s snapshots could create a new RBD cloned image (and then it follows the previous rule about 'move-to-trash' upon delete). 

[1] https://trello.com/c/JO5BPkRG

Reply:
https://github.com/ceph/ceph-csi/issues/424